boa_engine/src/object/internal_methods/mod.rs

```diff
@@ -294,14 +302,30 @@ pub(crate) struct InternalObjectMethods {
     pub(crate) __set_prototype_of__: fn(&JsObject, JsPrototype, &mut Context<'_>) -> JsResult<bool>,
     pub(crate) __is_extensible__: fn(&JsObject, &mut Context<'_>) -> JsResult<bool>,
     pub(crate) __prevent_extensions__: fn(&JsObject, &mut Context<'_>) -> JsResult<bool>,
-    pub(crate) __get_own_property__:
-        fn(&JsObject, &PropertyKey, &mut Context<'_>) -> JsResult<Option<PropertyDescriptor>>,
-    pub(crate) __define_own_property__:
-        fn(&JsObject, &PropertyKey, PropertyDescriptor, &mut Context<'_>) -> JsResult<bool>,
-    pub(crate) __has_property__: fn(&JsObject, &PropertyKey, &mut Context<'_>) -> JsResult<bool>,
-    pub(crate) __get__: fn(&JsObject, &PropertyKey, JsValue, &mut Context<'_>) -> JsResult<JsValue>,
+    pub(crate) __get_own_property__: fn(
+        &JsObject,
+        &PropertyKey,
+        &mut Slot,
+        &mut Context<'_>,
+    ) -> JsResult<Option<PropertyDescriptor>>,
+    pub(crate) __define_own_property__: fn(
+        &JsObject,
+        &PropertyKey,
+        PropertyDescriptor,
+        slot: &mut Slot,
+        &mut Context<'_>,
+    ) -> JsResult<bool>,
+    pub(crate) __has_property__:
+        fn(&JsObject, &PropertyKey, slot: &mut Slot, &mut Context<'_>) -> JsResult<bool>,
+    pub(crate) __get__: fn(
+        &JsObject,
+        &PropertyKey,
+        JsValue,
+        slot: &mut Slot,
+        &mut Context<'_>,
+    ) -> JsResult<JsValue>,
```

Comment:
I'm seeing a lot of usages of the `slot` argument is to pass a new `Slot` with a new `key`, which is why I think API-wise it'd be better to combine both into a `PropertyKey` that has a `slot: Cell<Slot>` property that is automatically initialized for current `Into<PropertyKey>` implementors.

Reply:
`slot` is an out parameter so it wouldn't work with having be merged into `PropertyKey` (we would have to change all call sites to take a reference).

And in the future we would have to pass more than just a slot, currently only direct object and it's `__proro__` properties are inline cached and a slot is sufficient for this.

We could just have methods on `JsObject` with non-sloted variant with just call internal object methods with `Slot::new()`. i.e `JsObject::__set__()` and `JsObject::__set_with_slot__` (or something like this)

Reply:
My main worry is that at one point we will have to expose the internal methods of an object, since the spec [requires allowing custom exotic objects for hosts](https://tc39.es/ecma262/#sec-host-internal-methods-of-exotic-objects), so I'd prefer to have those be as close to the spec as possible.

> `slot` is an out parameter so it wouldn't work with having be merged into `PropertyKey` (we would have to change all call sites to take a reference).

We already take a reference for the `PropertyKey` on all property-related operations, it's just that it's an immutable one. That's why I proposed putting a `Cell<Slot>` property on `PropertyKey`, since it makes a nicer API where you can do a simpler `"key".into()` if you're not interested in using the slot, but you can also pass a reference to a variable then inspect the slot after the operation using something like `prop.slot()`. 

Another option would be to take a `&mut PropertyKey` instead. However, it doesn't make much sense semantically to have a key be mutated by access operations, but the option is there.

> We could just have methods on `JsObject` with non-sloted variant with just call internal object methods with `Slot::new()`. i.e `JsObject::__set__()` and `JsObject::__set_with_slot__` (or something like this)

If you don't agree with my previous suggestion, I'd very much prefer this, since we can hide the custom internal methods and only leave the "spec related" ones public.

Reply:
If we have to expose the internal methods so the user override some/all, then how about this approach: We have a light struct that contains the `&mut Context` and any other internal crate data like `Slot`.

```rust
struct InternalMethodContext<'ctx> {
    context: &'ctx mut Context,
    slot: Slot,
    // any other internal method specific data.
}

impl<'ctx> InternalMethodContext<'ctx> {
   pub fn new(&'ctx mut context) -> Self {
       Self {
           context,
           slot: Slot::new,
       }
   }
   pub(crate) fn with_slot(slot: Slot) -> Self {
       Self {
           context,
           slot,
       }
   }
   pub(crate) fn slot(&mut self) -> &mut Slot {
       &mut self.slot
   }
}

impl Deref for InternalMethodContext {
    // returns context
}
// .. DerefMut
```

This would be a drop-in replacement for the `context` parameter. And where we call the internal methods we would do `__set__(..., &mut InternalMethodContext::new(context))` (the user only provides the methods we worry about the construction of the context anyway, so the user wouldn't even have to do this.)

This would also allow us to add or remove data more easily without making it a breaking change.